svf-llvm/lib/SVFIRBuilder.cpp

```diff
@@ -1207,92 +1170,313 @@ void SVFIRBuilder::addComplexConsForExt(const SVFValue* D, const SVFValue* S, co
-                        // Indirect call
-                        if (getCallee(svfcall) == nullptr)
-                            objId = pag->addDummyObjNode(svfcall->getType());
-                        else // Direct call
-                            objId = pag->getObjectNode(svfcall);
-
-                        operands.push_back(objId);
-                        nodeIDMap[s] = objId;
-                    }
-                }
-                // return value = -4 is a nullptr node
-                else if (nodeIDType == -4)
-                {
-                    operands.push_back(pag->getNullPtr());
-                    nodeIDMap[s] = operands[operands.size() - 1];
-                }
-                // return value = -5 is an offset
-                else if (nodeIDType == -5)
-                {
-                    for (char const &c : s)
-                    {
-                        if (std::isdigit(c) == 0)
-                            assert(false && "Invalid offset!");
-                    }
-                    operands.push_back(atoi(s.c_str()));
-                    nodeIDMap[s] = atoi(s.c_str());
-                }
-                // return value = -6 is an illegal operand format
-                else
-                    assert(false && "The operand format of function operation is illegal!");
+                NodeID srcId = getExtID(extOperation.GetBasicOp().GetType(), extOperation.GetBasicOp().GetSrcValue(), svfcall, extFunctionOps);
+                extOperation.GetBasicOp().SetSrcID(srcId);
+                nodeIDMap[s] = srcId;
+            }
+        }
+        if (!extOperation.GetBasicOp().GetDstValue().empty())
+        {
+            std::string s = extOperation.GetBasicOp().GetDstValue();
+            if (nodeIDMap.find(s) != nodeIDMap.end())
+                extOperation.GetBasicOp().SetDstID(nodeIDMap[s]);
+            else
+            {
+                NodeID dstId = getExtID(extOperation.GetBasicOp().GetType(), extOperation.GetBasicOp().GetDstValue(), svfcall, extFunctionOps);
+                extOperation.GetBasicOp().SetDstID(dstId);
+                nodeIDMap[s] = dstId;
+            }
+        }
+        // Get offset or size
+        if (!extOperation.GetBasicOp().GetOffsetOrSizeStr().empty())
+        {
+            std::string s = extOperation.GetBasicOp().GetOffsetOrSizeStr();
+            if (nodeIDMap.find(s) != nodeIDMap.end())
+                extOperation.GetBasicOp().SetOffsetOrSize(nodeIDMap[s]);
+            else
+            {
+                NodeID offsetOrSize = getExtID(extOperation.GetBasicOp().GetType(), extOperation.GetBasicOp().GetOffsetOrSizeStr(), svfcall, extFunctionOps);
+                extOperation.GetBasicOp().SetOffsetOrSize(offsetOrSize);
+                nodeIDMap[s] = offsetOrSize;
             }
         }
     }
 }
 
+SVFCallInst* SVFIRBuilder::addSVFExtCallInst(SVFBasicBlock* svfBB, const SVFFunction* svfcaller, const SVFFunction* svfcallee)
```

Comment:
One of my concerns is that Codecov reported that code added in this file was not covered by tests. It would be good if we can add a few test cases.